src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,167 @@
+      val info = createAnnotatedInfo(r.info)
+      val assertBounds = Assert(assertionExprInBounds(r, l.op.typ)(info))(info)
+      Seqn(Vector(assertBounds, ass))(l.op.info)
+
+    case f@FunctionCall(_, _, args) =>
+      val asserts = args map (expr => {
+        val info = createAnnotatedInfo(expr.info)
+        Assert(assertionExprInBounds(expr, expr.typ)(info))(info)
+      })
+      Seqn(f +: asserts)(f.info)
+
+    case m@MethodCall(_, recv, _, args) =>
+      val asserts = args map (expr => {
+        val info = createAnnotatedInfo(expr.info)
+        Assert(assertionExprInBounds(expr, expr.typ)(info))(info)
+      })
+      val recvInfo = createAnnotatedInfo(recv.info)
+      val recvAssert = Assert(assertionExprInBounds(recv, recv.typ)(recvInfo))(recvInfo)
+      Seqn(m +: recvAssert +: asserts)(m.info)
+
+    case m@Make(_, typ) =>
+      val info = createAnnotatedInfo(typ.info)
+      val assertBounds = Assert(assertionExprInBounds(typ.op, typ.op.typ)(info))(info)
+      Seqn(Vector(assertBounds, m))(m.info)
+
+    // explicitly matches remaining statements to detect non-exhaustive pattern matching if a new statement is added
+    case x@(_: Inhale | _: Exhale | _: Assert | _: Assume | _: Return | _: Fold | _: Unfold) => x
+  }
+
+  // Checks if expr and its subexpressions are within bounds of type `typ`
+  private def assertionExprInBounds(expr: Expr, typ: Type)(info: Source.Parser.Info): Assertion = {
+    val trueLit: Expr = BoolLit(true)(info)
+
+    def genAssertionExpr(expr: Expr, typ: Type): Expr =
+      typ match {
+        case IntT(_, kind) if kind.isInstanceOf[BoundedIntegerKind] =>
+          val boundedKind = kind.asInstanceOf[BoundedIntegerKind]
+          And(
+            AtMostCmp(IntLit(boundedKind.lower)(info), expr)(info),
+            AtMostCmp(expr, IntLit(boundedKind.upper)(info))(info))(info)
+
+        case _ => trueLit
+      }
+
+    val intSubExprs: Set[Expr] = Expr.getSubExpressions(expr).filter(_.typ.isInstanceOf[IntT])
+
+    // values assumed to be within bounds, i.e. variables, fields from structs, dereferences of pointers and indexed expressions
+    // this stops Gobra from throwing overflow errors in field accesses and pointer dereferences because Gobra was not able to prove that
+    // they were within bounds even though that is guaranteed by the expression's type
+    val valuesWithinBounds = intSubExprs.filter {
+      case _: Var | _: FieldRef | _: IndexedExp | _: Deref => true
+      case _ => false
+    }
+
+    val computeAssertions = (exprs: Set[Expr]) => exprs.map{x => genAssertionExpr(x, x.typ)}.foldRight(trueLit)((x,y) => And(x,y)(info))
+
+    // assumptions for the values that are considered within bounds
+    val assumptions = computeAssertions(valuesWithinBounds)
+
+    // Assertions that need to be verified assuming the expressions in `assumptions`
+    val obligations = ExprAssertion(computeAssertions(intSubExprs))(info)
+    Implication(assumptions, obligations)(info)
```

Comment:
The assumptions about the values of field reads is only added here, right? But if it is only added as an asumption, then what happens if the assumption comes from a different line? e.g. `x := y.f; return x`

Reply:
This is done not only for field reads but also for variables, indexed expressions and pointer dereferences. As such, the value of a variable will always be assumed to be within bounds (my point here is that when the result of an operation that overflows is stored in a variable, we don't need to check for overflows in posterior uses of the variable because we check for overflows when we do the assignment).